src/plugin/declarativedbusabstractobject.cpp

```diff
@@ -0,0 +1,245 @@
+#include <QDBusMessage>
+#include <QDBusConnection>
+
+#include <QTimerEvent>
+
+#include <qqmlinfo.h>
+
+DeclarativeDBusAbstractObject::DeclarativeDBusAbstractObject(QObject *parent)
+    : QDBusVirtualObject(parent)
+    , m_bus(DeclarativeDBus::SessionBus)
+    , m_quitTimerId(-1)
+    , m_quitTimeout(30)
+    , m_quitOnTimeout(false)
+    , m_complete(true)
+{
+}
+
+DeclarativeDBusAbstractObject::~DeclarativeDBusAbstractObject()
+{
+    QDBusConnection conn = DeclarativeDBus::connection(m_bus);
+    conn.unregisterObject(m_path);
+
+    // In theory an application could have multiple AbstractObject items for different paths or
+    // interfaces and destroying one would unregister the whole service.
+    // If problems arise we could introduce refcounting for services so that
+    // the last service would be responsible for unregisting the service upon
+    // destruction. Unregisration should also take into account unregistration happening
+    // from C++ side.
+    if (!m_service.isEmpty()) {
+        if (!conn.unregisterService(m_service)) {
+            qmlInfo(this) << "Failed to unregister service " << m_service;
+            qmlInfo(this) << conn.lastError().message();
+        }
+    }
+}
+
+QString DeclarativeDBusAbstractObject::service() const
+{
+    return m_service;
+}
+
+void DeclarativeDBusAbstractObject::setService(const QString &service)
+{
+    if (m_service != service) {
+        m_service = service;
+        emit serviceChanged();
+    }
+}
+
+QString DeclarativeDBusAbstractObject::path() const
+{
+    return m_path;
+}
+
+void DeclarativeDBusAbstractObject::setPath(const QString &path)
+{
+    if (m_path != path) {
+        m_path = path;
+        emit pathChanged();
+    }
+}
+
+/*
+    The XML service description. This could be derived from the meta-object but since it's unlikely
+    to be needed most of the time this and type conversion is non-trivial it's not, and there is
+    this property instead if it is needed.
+*/
+
+QString DeclarativeDBusAbstractObject::xml() const
+{
+    return m_xml;
+}
+
+void DeclarativeDBusAbstractObject::setXml(const QString &xml)
+{
+    if (m_xml != xml) {
+        m_xml = xml;
+        emit xmlChanged();
+    }
+}
+
+DeclarativeDBus::BusType DeclarativeDBusAbstractObject::bus() const
+{
+    return m_bus;
+}
+
+void DeclarativeDBusAbstractObject::setBus(DeclarativeDBus::BusType bus)
+{
+    if (m_bus != bus) {
+        m_bus = bus;
+        emit busChanged();
+    }
+}
+
+int DeclarativeDBusAbstractObject::quitTimeout() const
+{
+    return m_quitTimeout;
+}
+
+void DeclarativeDBusAbstractObject::setQuitTimeout(int timeout)
+{
+    if (m_quitTimeout != timeout) {
+        m_quitTimeout = timeout;
+
+        emit quitTimeoutChanged();
+    }
+}
+
+
+bool DeclarativeDBusAbstractObject::quitOnTimeout() const
+{
+    return m_quitOnTimeout;
+}
+
+void DeclarativeDBusAbstractObject::setQuitOnTimeout(bool quit)
+{
+    if (m_quitOnTimeout != quit && (!m_complete || !quit)) {
```

Comment:
Yes, the app won't be active initially so auto quit is enabled on component complete, when the application becomes active later it be disabled cancelling the timeout and it can't be turned on again.

In theory, on the application becoming active there are guaranteed visible windows and removing the event loop locker won't exit the application. And it's less laborious than explicitly cancelling the quit on every method call that results in a window being shown, and in the case where the application isn't started by dbus. I would have preferred to use ApplicationWindow.visible but that was true initially despite item visibility following window visibility and the window not yet having been shown.

Unfortunately I forgot about starting a second application while a first is still launching, in which case the first application will never hit that application.active state despite having shown a window and it will exit after a bit. So that's no good either, and we're trending back to so convoluted to use that you're better off rolling your own so you can at least follow what's happening territory.